### PR TITLE
test: Update Python interop tests to python-tuf v1.0.0

### LINKED ISSUE #221: interop tests use deprecated python-tuf code
client/python_interop uses the legacy python-tuf client code (the generate.py script also uses legacy repository_tool). python-tuf has recently removed the legacy code from git, and the next release will not contain it anymore.

Updating to new client should be straight forward: There's an example client that might be what you need (https://github.com/theupdateframework/python-tuf/tree/develop/examples/client_example) -- I admit I did not fully find out what the test does.

Updating generate.py might be a bit more work (in the sense that there is no repository_tool API anymore) but amount of code should not be any worse using the low level Metadata API, see https://github.com/theupdateframework/python-tuf/tree/develop/examples/repo_example/basic_repo.py

python-tuf devs are obviously available for advice.

### PR DESCRIPTION
**Pull request by znewman01**
Fixes #221

### CODE REVIEW

**Review by jku — COMMENTED** [review]
Left some comments hopefully they are helpful

The metadata generation is quite verbose as is visible here: I hope we can improve on that soon, I'll try to remember to send a patch here when that happens.

> there's no built-in support for consistent snapshots, right?

I think the metadata you now produce is completely valid consistent_snapshot (the non-consistent version seems a bit incorrect as root is not versioned but I commented on that in code). Let me know if there's anything that seems missing.

It's true we don't currently handle target files themselves in any way so you will need to produce the hash prefixed filename yourself when saving the file, that is a bit annoying: I left a comment in the code.

> is there a better way to make sure that timestamp.json has correct hashes/length than what I'm doing?

this is an oversight in python-tuf: we've not been using the optional hashes in meta dicts

> TAP-12 seemingly got implemented in python-tuf v1.0.0

Maybe more accurate to say Metadata API does not enforce or test that the id actually is the hex digest of the canonical form: it's possible that as an accidental result of that we are producing metadata that is not strictly conformant with that current requirement.

File: client/python_interop/testdata/python-tuf-v1.0.0/generate.py
Comment (znewman01) [reply client/python_interop/testdata/python-tuf-v1.0.0/generate.py:124]: go-tuf barfs if these are omitted; filed #229 

(though only for timestamp meta, not snapshot meta)

File: client/python_interop/testdata/python-tuf-v1.0.0/generate.py
Comment (znewman01) [reply client/python_interop/testdata/python-tuf-v1.0.0/generate.py]: ACK fixed

File: client/python_interop/testdata/python-tuf-v1.0.0/generate.py
Comment (znewman01) [reply client/python_interop/testdata/python-tuf-v1.0.0/generate.py:44]: I omitted this before because it feels like I'm testing *my* consistent snapshot implementation, rather than the `python-tuf` consistent snapshot implementation. I guess it makes sense to put it in for now and then swap out when the APIs have changed.

**Comment by znewman01** [comment]
@jku:

> Left some comments hopefully they are helpful

Yes, thank you for the quick review :smile: great context

> Maybe more accurate to say Metadata API does not enforce or test that the id actually is the hex digest of the canonical form: it's possible that as an accidental result of that we are producing metadata that is not strictly conformant with that current requirement.

ACK, makes sense. I will consider this a python-tuf bug and track [python-tuf#1894](https://github.com/theupdateframework/python-tuf/issues/1894).

However, it seems like it should be okay to ignore the keyIDs as a client and barring objections I will update the PR to allow this.

File: client/python_interop/testdata/python-tuf-v1.0.0/generate.py
Comment (jku) [reply client/python_interop/testdata/python-tuf-v1.0.0/generate.py:44]: We only have an API for modifying metadata: serving the target files is not part of the equation in any way.

This is likely to improve in future somehow but I think it's unlikely we'd add this functionality to Metadata API itself.

Comment (znewman01) [reply client/python_interop/testdata/python-tuf-v1.0.0/generate.py:44]: ACK that makes sense :)

File: verify/db.go
Comment (znewman01) [reply verify/db.go:67]: Thanks!

**Comment by mnm678** [comment]
> > TAP-12 seemingly got implemented in python-tuf v1.0.0
> 
> Maybe more accurate to say Metadata API does not enforce or test that the id actually is the hex digest of the canonical form: it's possible that as an accidental result of that we are producing metadata that is not strictly conformant with that current requirement.

With the python-tuf implementation, I think that we can move TAP 12 to accepted pretty quickly. Unless there are strong objections to the TAP that will lead to major changes (which I haven't seen), I'd argue for future-proofing go-tuf by implementing TAP 12 now. It may make the implementation temporarily out of sync with the specification, but not in a major way, and not in a way that affects security (as described in more detail in [the TAP](https://github.com/theupdateframework/taps/blob/master/tap12.md)).

**Comment by znewman01** [comment]
Filed #232 for TAP-12.

I think we should probably block this PR on that issue. That would make sure we don't half-implement TAP-12 and accidentally fail-open.

**Comment by joshuagl** [comment]
As #232 is merged can we rebase this and get it reviewed @znewman01?

**Comment by znewman01** [comment]
Yup! Might be a week-ish until I get around to it.

**Comment by znewman01** [comment]
Go for it! Let me know if you don’t find time,it had finally popped to the top of my TODO list this week but happy to have you take it off my hands

**Review by asraa — COMMENTED** [review]
@znewman01 rebased, tests passing

File: client/python_interop/testdata/python-tuf-v1.0.0/generate.py
Comment (asraa) [reply client/python_interop/testdata/python-tuf-v1.0.0/generate.py:124]: Since this is resolved now, I removed this.

**Comment by asraa** [comment]
@jku @mnm678 could you PTAL? TAP-12 key ID removal is done, and so is some minor go-tuf snapshot length/hashes presence bug is resolved too.

**Review by mnm678 — APPROVED** [review]

**Comment by znewman01** [comment]
LGTM though I can't review it since it's my PR

**Comment by trishankatdatadog** [comment]
I'll leave it to @jku since he's best suited to review this one

**Comment by jku** [comment]
> I'll leave it to @jku since he's best suited to review this one

Sorry, I'm unlikely to have a real look in then next weeks.

File: client/python_interop/testdata/python-tuf-v1.0.0/generate.py
Comment (trishankatdatadog) [inline client/python_interop/testdata/python-tuf-v1.0.0/generate.py:77]: We may want to include additional, optional metadata such as hashes and length, but good for now.

**Review by trishankatdatadog — APPROVED** [review]

**Comment by trishankatdatadog** [comment]
I made an [exception](https://github.com/theupdateframework/go-tuf/pull/228/checks?check_run_id=8041608265) for the DCO this time.

**Pull request merged**
